lib/bap_image/llvm_binary_stubs.cpp

```diff
@@ -0,0 +1,172 @@
+#include <memory>
+#include <string>
+
+#include "llvm_binary_stubs.h"
+#include "llvm_binary_stubs.hpp"
+
+#include <llvm/Support/MemoryBuffer.h>
+#include <llvm/Object/Binary.h>
+#include <llvm/Support/system_error.h>
+#include <llvm/ADT/Triple.h>
+#include <llvm/Object/Archive.h>
+#include <llvm/Object/ObjectFile.h>
+#include <llvm/Support/Casting.h>
+
+#include "llvm_binary_stubs.h"
+#include "llvm_segment.hpp"
+
+extern "C" {
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/fail.h>
+#include <caml/alloc.h>
+#include <caml/custom.h>
+#include <caml/bigarray.h>
+}
+
+void llvm_binary_fail [[ noreturn ]](const char* message) {
+    ::caml_failwith(message);
+}
+
+namespace impl {
+
+using namespace llvm;
+using namespace llvm::object;
+
+Binary* from_value (::value v) {
+    return *reinterpret_cast<Binary**>(Data_custom_val(v));
+}
+
+::value to_value (Binary* b) {
+    static struct custom_operations binary_ops = {
+        const_cast<char*>("llvm.caml.llvm_binary"),
+        [](::value v) { delete from_value(v); },
+        custom_compare_default,
+        custom_hash_default,
+        custom_serialize_default,
+        custom_deserialize_default
+    };
+    ::value v = ::alloc_custom(&binary_ops, sizeof(Binary*), 0, 1);
+    *reinterpret_cast<Binary**>(Data_custom_val(v)) = b;
+    return v;
+}
+
+
+::value to_value(Triple::ArchType arch) {
+    typedef Triple::ArchType AT;
+    switch(arch) {
+    case AT::arm: return ::caml_hash_variant("arm");
+    case AT::aarch64: return ::caml_hash_variant("aarch64");
+    case AT::hexagon: return ::caml_hash_variant("hexagon");
+    case AT::mips: return ::caml_hash_variant("mips");
+    case AT::mipsel: return ::caml_hash_variant("mipsel");
+    case AT::mips64: return ::caml_hash_variant("mips64");
+    case AT::mips64el: return ::caml_hash_variant("mips64el");
+    case AT::ppc: return ::caml_hash_variant("ppc");
+    case AT::ppc64: return ::caml_hash_variant("ppc64");
+    case AT::ppc64le: return ::caml_hash_variant("ppc64le");
+    case AT::r600: return ::caml_hash_variant("r600");
+    case AT::sparc: return ::caml_hash_variant("sparc");
+    case AT::sparcv9: return ::caml_hash_variant("sparcv9");
+    case AT::systemz: return ::caml_hash_variant("systemz");
+    case AT::thumb: return ::caml_hash_variant("thumb");
+    case AT::x86: return ::caml_hash_variant("x86");
+    case AT::x86_64: return ::caml_hash_variant("x86_64");
+    case AT::xcore: return ::caml_hash_variant("xcore");
+    case AT::nvptx: return ::caml_hash_variant("nvptx");
+    case AT::nvptx64: return ::caml_hash_variant("nvptx64");
+    default:
+        std::string message("Unknown arch: ");
+        message += std::to_string(static_cast<int>(arch));
+        ::caml_failwith(message.c_str());
+    }
+}
+
+
+::value to_value(const seg::segment& s) {
```

Comment:
All allocating function must start with `CAMLparam*` macro:

   > The function body must start with one of the [CAMLparam] macros.
   If the function has no parameter of type [value], use [CAMLparam0].
   If the function has 1 to 5 [value] parameters, use the corresponding
   [CAMLparam] with the parameters as arguments.
   If the function has more than 5 [value] parameters, use [CAMLparam5]
   for the first 5 parameters, and one or more calls to the [CAMLxparam]
   macros for the others.
   If the function takes an array of [value]s as argument, use
   [CAMLparamN] to declare it (or [CAMLxparamN] if you already have a
   call to [CAMLparam] for some other arguments).

   > If you need local variables of type [value], declare them with one
   or more calls to the [CAMLlocal] macros at the beginning of the
   function, after the call to CAMLparam.  Use [CAMLlocalN] (at the
   beginning of the function) to declare an array of [value]s.